src/Microsoft.Extensions.WebEncoders.Core/UrlPathDecoder.cs

```diff
@@ -0,0 +1,326 @@
+                {
+                    output = new char[] { first };
+                }
+                else
+                {
+                    output[0] = first;
+                }
+
+                count = 1;
+                return 3;
+            }
+
+            // anticipate the byte count
+            int currentDecodeBits = 0;
+            int bytesCount = 1;
+            int expectValueMin = 0;
+            bytesBuffer[0] = (byte)first;
+            if ((first & 0xE0) == 0xC0)
+            {
+                // 110x xxxx, expect 1 more byte
+                currentDecodeBits = first & 0x1F;
+                bytesCount = 2;
+                expectValueMin = 0x80;
+            }
+            else if ((first & 0xF0) == 0xE0)
+            {
+                // 1110 xxxx, expect 2 more bytes
+                currentDecodeBits = first & 0x0F;
+                bytesCount = 3;
+                expectValueMin = 0x800;
+            }
+            else if ((first & 0xF8) == 0xF0)
+            {
+                // 1111 0xxx, expect 3 more bytes
+                currentDecodeBits = first & 0x07;
+                bytesCount = 4;
+                expectValueMin = 0x10000;
+            }
+            else
+            {
+                // invalid
+                return 0;
+            }
+
+            if (start + (bytesCount * 3) > source.Length)
+            {
+                // less than expected bytes to decode
+                return 0;
+            }
+
+            var remainingBytes = bytesCount - 1;
+            while (remainingBytes > 0)
+            {
+                start += 3;
+                if (source[start] != '%')
+                {
+                    return 0;
+                }
+
+                char v;
+                if (!TryGetUnescapedAscii(source[start + 1], source[start + 2], out v))
+                {
+                    return 0;
+                }
+
+                if ((v & 0xC0) != 0x80)
+                {
+                    return 0;
+                }
+
+                // append the lower 6bit
+                currentDecodeBits = (currentDecodeBits << 6) | (v & 0x3F);
+
+                bytesBuffer[bytesCount - remainingBytes] = (byte)v;
+                remainingBytes--;
+
+                if (remainingBytes == 1 && currentDecodeBits >= 0x360 && currentDecodeBits <= 0x37F)
+                {
+                    // This is going to end up in the range of 0xD800-0xDFFF UTF-16 surrogates that are not allowed in UTF-8;
+                    return 0;
+                }
+                if (remainingBytes == 2 && currentDecodeBits >= 0x110)
+                {
+                    // This is going to be out of the upper Unicode bound 0x10FFFF.
+                    return 0;
+                }
+            }
+
+            if (currentDecodeBits < expectValueMin)
+            {
+                // Overlong encoding (e.g. using 2 bytes to encode something that only needed 1).
```

Comment:
I think this a place where normalization would be performed.
https://tools.ietf.org/html/rfc3987#section-5.3.2.2